sw/src/offline-interface/offline-interface.js

```diff
@@ -0,0 +1,191 @@
+     * @param {String} options.sectionId - ID of section to record
+     * @param {Number} options.recordingTimeoutDelay - How long to wait after all pending requests have resolved before stopping recording, if no other requests come in
+     * @param {Function} options.onStarted - Will be called when the service worker is set up for recording and ready to go.
+     * @param {Function} options.onCompleted - Called when recording completes successfully and the cached section has been saved in the IndexedDB and CacheStorage.
+     * @param {Function} options.onError - Called if there's an error during recording; receives an Error object as an argument.
+     */
+    async startRecording({
+        sectionId,
+        recordingTimeoutDelay,
+        onStarted,
+        onCompleted,
+        onError,
+    }) {
+        if (!this.initialized)
+            throw new Error(
+                'OfflineInterface has not been initialized. Make sure `pwa.enabled` is `true` in `d2.config.js`'
+            )
+        if (!sectionId || !onStarted || !onCompleted || !onError)
+            throw new Error(
+                '[Offline interface] The options { sectionId, onStarted, onCompleted, onError } are required when calling startRecording()'
+            )
+
+        // Send SW message to start recording
+        swMessage(swMsgs.startRecording, {
+            sectionId,
+            recordingTimeoutDelay,
+        })
+
+        /** Cleans up SW recording listeners */
+        const cleanUpListeners = () => {
+            const messageTypes = [
+                swMsgs.recordingStarted,
+                swMsgs.recordingError,
+                swMsgs.recordingCompleted,
+                swMsgs.confirmRecordingCompletion,
+            ]
+            messageTypes.forEach(messageType =>
+                this.offlineEvents.removeAllListeners(messageType)
+            )
+        }
+
+        // Prep for subsequent events after recording starts:
+        this.offlineEvents.once(swMsgs.recordingStarted, onStarted)
+        this.offlineEvents.once(
+            swMsgs.confirmRecordingCompletion,
+            // Confirms recording is okay to save
+            () => swMessage(swMsgs.completeRecording)
+        )
+        this.offlineEvents.once(swMsgs.recordingCompleted, () => {
+            cleanUpListeners()
+            onCompleted()
+        })
+        this.offlineEvents.once(swMsgs.recordingError, ({ error }) => {
+            cleanUpListeners()
+            onError(error)
+        })
+    }
+
+    /**
+     * Retrieves a list of cached sections from IndexedDB. Creates DB if it
+     * doesn't exist yet to avoid race conditions with service worker.
+     * @returns {Promise} A promise that resolves to an array of cached sections.
+     */
+    async getCachedSections() {
+        // This fn may be called before service worker is ready. Wait 'til then
+        // to avoid opening a DB in an app that isn't PWA-enabled
+        await navigator.serviceWorker.ready
```

Comment:
This was a naive way to handle this function if PWA is not enabled - I will refactor this